script/bulk_import/generic_bulk.rb

```diff
@@ -2359,6 +2420,455 @@ def import_permalinks
+      next if !chatable_id
+      # TODO: Add more uniqueness checks
+      #       Ensure no channel with same name and category exists?
+
+      {
+        original_id: row["id"],
+        name: row["name"],
+        description: row["description"],
+        slug: row["slug"],
+        status: row["status"],
+        chatable_id: chatable_id,
+        chatable_type: row["chatable_type"],
+        user_count: row["user_count"],
+        messages_count: row["messages_count"],
+        type: type,
+        created_at: to_datetime(row["created_at"]),
+        allow_channel_wide_mentions: to_boolean(row["allow_channel_wide_mentions"]),
+        auto_join_users: to_boolean(row["auto_join_users"]),
+        threading_enabled: to_boolean(row["threading_enabled"]),
+      }
+    end
+
+    channels.close
+  end
+
+  def import_user_chat_channel_memberships
+    unless defined?(::Chat)
+      puts "", "Skipping user chat channel memberships, because the chat plugin is not installed."
+      return
+    end
+
+    puts "", "Importing user chat channel memberships..."
+
+    channel_users = query(<<~SQL)
+      SELECT chat_channels.chatable_type, chat_channels.chatable_id, chat_channel_users.*
+        FROM chat_channel_users
+             JOIN chat_channels ON chat_channels.id = chat_channel_users.chat_channel_id
+       ORDER BY chat_channel_users.chat_channel_id
+    SQL
+
+    existing_members =
+      Chat::UserChatChannelMembership.distinct.pluck(:user_id, :chat_channel_id).to_set
+
+    create_user_chat_channel_memberships(channel_users) do |row|
+      user_id = user_id_from_imported_id(row["user_id"])
+      channel_id = chat_channel_id_from_original_id(row["chat_channel_id"])
+      last_read_message_id = chat_message_id_from_original_id(row["last_read_message_id"])
+
+      next if user_id.blank? || channel_id.blank?
+      next unless existing_members.add?([user_id, channel_id])
+
+      # `last_viewed_at` is required, if not provided, set a placeholder,
+      # it'll be updated in the `update_chat_membership_metadata` step
+      last_viewed_at = to_datetime(row["last_viewed_at"].presence || LAST_VIEWED_AT_PLACEHOLDER)
+
+      {
+        user_id: user_id,
+        chat_channel_id: channel_id,
+        created_at: to_datetime(row["created_at"]),
+        following: to_boolean(row["following"]),
+        muted: to_boolean(row["muted"]),
+        desktop_notification_level: row["desktop_notification_level"],
+        mobile_notification_level: row["mobile_notification_level"],
+        last_read_message_id: last_read_message_id,
+        join_mode: row["join_mode"],
+        last_viewed_at: last_viewed_at,
+      }
+    end
+
+    puts "", "Importing chat direct message users..."
+
+    channel_users.reset
+    existing_direct_message_users =
+      Chat::DirectMessageUser.distinct.pluck(:direct_message_channel_id, :user_id).to_set
+
+    create_direct_message_users(channel_users) do |row|
+      next if row["chatable_type"] != "DirectMessage"
```

Comment:
Wouldn't it be easier and maybe faster to use separate SQL and include the condition in SQL?

Reply:
It might be. I’ll look into it during the port to the new framework.  I just assumed that `channel_users.reset` was doing some type of lightweight "cursor" reset which should be faster than a brand  new query.